# Turbofan Engines Predictive Maintenance

## EDA Phase

""" Aircraft Turbofan Engines Predictive Maintenance:
This notebook is the EDA phase for 11 turbofan Engines,
the ultimate goal of the complete project is to predict
the remaining useful life (RUL) of similar in-service engines """


In [2]:
# Import neccessary libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd


In [3]:
# load the data
df = pd.read_csv('engines data 130325 masked.csv') 


In [4]:
df.head(20)

,EPOSITION,FLIGHT_PHASE,FLIGHT_DATETIME,ENGINE_INSTALL_DATE,Cycles,EGT Probe average,Fuel Flow,Corrected Fan Speed,Core Speed,ZPN12P,Vibration N1 #1 Bearing,Vibration N2 #1 Bearing,Vibration N2 Turbine Frame,Engine Type,Engine Number
0,1,CRUISE,29-11-23 8:56,24-08-23 23:00,334.0,793.3125,2782.0,92.2129,103.12,86.67,0.1267,0.500,0.300,1,11
1,1,CRUISE,07-09-23 1:10,24-08-23 23:00,32.0,792.9250,2725.0,90.0819,103.99,86.02,0.1200,0.200,0.375,1,11
2,1,CRUISE,19-01-24 3:09,24-08-23 23:00,532.0,757.2250,2861.0,89.0560,101.87,83.67,0.0933,0.350,0.600,1,11
3,1,CLIMB,25-02-24 21:58,24-08-23 23:00,675.0,905.0375,5959.0,NaN,107.42,92.15,0.1133,0.450,0.475,1,11
4,1,TAKEOFF,10-07-24 16:28,24-08-23 23:00,1223.0,912.3625,2911.0,85.3176,106.87,86.45,0.0600,0.850,0.900,1,11
5,1,CRUISE,23-02-25 23:45,24-08-23 23:00,126.0,806.4375,2650.0,92.0863,103.02,86.50,0.0000,0.700,0.600,1,11
6,1,CRUISE,17-03-24 8:01,24-08-23 23:00,766.0,715.6875,2504.0,88.0891,99.87,81.40,0.0400,0.125,0.200,1,11
7,1,CRUISE,08-12-24 12:53,24-08-23 23:00,118.0,786.9625,3005.0,92.2527,102.16,85.80,0.0000,0.500,0.625,1,11
8,1,TAKEOFF,09-08-24 2:11,24-08-23 23:00,39.0,983.2375,4066.0,86.9904,110.24,90.41,0.0933,0.400,0.650,1,11
9,1,TAKEOFF,30-06-24 9:25,24-08-23 23:00,1178.0,938.3250,3206.0,85.0595,108.41,87.59,0.0667,0.650,0.675,1,11


In [5]:
# Check the data statistics
df.describe()

,EPOSITION,Cycles,EGT Probe average,Fuel Flow,Corrected Fan Speed,Core Speed,ZPN12P,Vibration N1 #1 Bearing,Vibration N2 #1 Bearing,Vibration N2 Turbine Frame,Engine Type,Engine Number
count,120677.000000,120672.000000,120610.000000,120603.000000,79972.000000,120643.000000,120643.000000,120379.000000,120377.000000,120379.000000,120677.000000,120677.000000
mean,1.513594,1919.581867,825.060648,3306.745645,84.621995,103.296047,83.395046,0.134921,0.339130,0.597473,1.782005,5.172319
std,0.499817,1388.496433,80.575206,1548.253271,5.013564,4.532367,8.490425,0.097757,0.240131,0.392315,0.412886,2.938684
min,1.000000,0.000000,0.000000,553.000000,68.178800,0.820000,0.240000,0.000000,0.000000,0.000000,1.000000,1.000000
25%,1.000000,682.000000,763.312500,2168.000000,79.768825,101.410000,81.060000,0.060000,0.150000,0.325000,2.000000,3.000000
50%,2.000000,1693.000000,849.937500,2704.000000,85.368350,104.280000,83.770000,0.113300,0.300000,0.500000,2.000000,5.000000
75%,2.000000,3076.000000,883.662500,5107.000000,89.027550,106.040000,88.330000,0.200000,0.475000,0.775000,2.000000,7.000000
max,2.000000,5003.000000,1047.600000,9925.000000,100.156900,112.130000,95.310000,0.793300,1.600000,2.525000,2.000000,11.000000


In [6]:
# check data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120677 entries, 0 to 120676
Data columns (total 15 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   EPOSITION                   120677 non-null  int64  
 1   FLIGHT_PHASE                120677 non-null  object 
 2   FLIGHT_DATETIME             120677 non-null  object 
 3   ENGINE_INSTALL_DATE         120677 non-null  object 
 4   Cycles                      120672 non-null  float64
 5   EGT Probe average           120610 non-null  float64
 6   Fuel Flow                   120603 non-null  float64
 7   Corrected Fan Speed         79972 non-null   float64
 8   Core Speed                  120643 non-null  float64
 9   ZPN12P                      120643 non-null  float64
 10  Vibration N1 #1 Bearing     120379 non-null  float64
 11  Vibration N2 #1 Bearing     120377 non-null  float64
 12  Vibration N2 Turbine Frame  120379 non-null  float64
 13  Engine Type   

In [7]:
# Check for missing values, percentage
df.isnull().sum()/len(df)*100


EPOSITION                      0.000000
FLIGHT_PHASE                   0.000000
FLIGHT_DATETIME                0.000000
ENGINE_INSTALL_DATE            0.000000
Cycles                         0.004143
EGT Probe average              0.055520
Fuel Flow                      0.061321
Corrected Fan Speed           33.730537
Core Speed                     0.028174
ZPN12P                         0.028174
Vibration N1 #1 Bearing        0.246940
Vibration N2 #1 Bearing        0.248597
Vibration N2 Turbine Frame     0.246940
Engine Type                    0.000000
Engine Number                  0.000000
dtype: float64

In [8]:
# delete features that have more than 50% missing values
null_50 = df.isnull().sum()/len(df)*100 > 50
null_50 = null_50[null_50].index
df = df.drop(null_50, axis=1)
df.head(20)
 

,EPOSITION,FLIGHT_PHASE,FLIGHT_DATETIME,ENGINE_INSTALL_DATE,Cycles,EGT Probe average,Fuel Flow,Corrected Fan Speed,Core Speed,ZPN12P,Vibration N1 #1 Bearing,Vibration N2 #1 Bearing,Vibration N2 Turbine Frame,Engine Type,Engine Number
0,1,CRUISE,29-11-23 8:56,24-08-23 23:00,334.0,793.3125,2782.0,92.2129,103.12,86.67,0.1267,0.500,0.300,1,11
1,1,CRUISE,07-09-23 1:10,24-08-23 23:00,32.0,792.9250,2725.0,90.0819,103.99,86.02,0.1200,0.200,0.375,1,11
2,1,CRUISE,19-01-24 3:09,24-08-23 23:00,532.0,757.2250,2861.0,89.0560,101.87,83.67,0.0933,0.350,0.600,1,11
3,1,CLIMB,25-02-24 21:58,24-08-23 23:00,675.0,905.0375,5959.0,NaN,107.42,92.15,0.1133,0.450,0.475,1,11
4,1,TAKEOFF,10-07-24 16:28,24-08-23 23:00,1223.0,912.3625,2911.0,85.3176,106.87,86.45,0.0600,0.850,0.900,1,11
5,1,CRUISE,23-02-25 23:45,24-08-23 23:00,126.0,806.4375,2650.0,92.0863,103.02,86.50,0.0000,0.700,0.600,1,11
6,1,CRUISE,17-03-24 8:01,24-08-23 23:00,766.0,715.6875,2504.0,88.0891,99.87,81.40,0.0400,0.125,0.200,1,11
7,1,CRUISE,08-12-24 12:53,24-08-23 23:00,118.0,786.9625,3005.0,92.2527,102.16,85.80,0.0000,0.500,0.625,1,11
8,1,TAKEOFF,09-08-24 2:11,24-08-23 23:00,39.0,983.2375,4066.0,86.9904,110.24,90.41,0.0933,0.400,0.650,1,11
9,1,TAKEOFF,30-06-24 9:25,24-08-23 23:00,1178.0,938.3250,3206.0,85.0595,108.41,87.59,0.0667,0.650,0.675,1,11


In [9]:
# lets rename the columns for easy access, replace space with underscore, and convert to lower case, and remove special characters, and remove the word 'engine', and make shorter names
df.columns = df.columns.str.replace(' ', '_').str.lower().str.replace('[-/()#]', '').str.replace('engine', 'eng').str.replace('temperature', 'temp').str.replace('pressure', 'pres').str.replace('vibration', 'vib').str.replace('speed', 'spd').str.replace('temperature', 'temp').str.replace('flow', 'flw')

In [10]:
df.columns

Index(['eposition', 'flight_phase', 'flight_datetime', 'eng_install_date',
       'cycles', 'egt_probe_average', 'fuel_flw', 'corrected_fan_spd',
       'core_spd', 'zpn12p', 'vib_n1_#1_bearing', 'vib_n2_#1_bearing',
       'vib_n2_turbine_frame', 'eng_type', 'eng_number'],
      dtype='object')

In [11]:
# check how many engines in the data
df['eng_number'].unique()

array([11,  8,  6,  7,  2,  5,  3, 10,  9,  1,  4], dtype=int64)

In [12]:
# make a dictionary of engines types and numbers for each type
eng_types = df['eng_type'].unique()
eng_numbers = {}
for eng in eng_types:
    eng_numbers[eng] = df[df['eng_type'] == eng]['eng_number'].unique()
eng_numbers


{1: array([11,  8, 10,  9], dtype=int64),
 2: array([6, 7, 2, 5, 3, 1, 4], dtype=int64)}

In [13]:
# Change the categorical dat types into suitable type

df['eng_number'] = df['eng_number'].astype('category')
df['eng_type'] = df['eng_type'].astype('category')
df['flight_phase'] = df['flight_phase'].astype('category')


In [14]:
# check the unique flight phases
df['flight_phase'].unique()

['CRUISE', 'CLIMB', 'TAKEOFF']
Categories (3, object): ['CLIMB', 'CRUISE', 'TAKEOFF']

In [15]:
# how many nulls in each flight phase
df.groupby('flight_phase').count()

C:\Users\Ahmad\AppData\Local\Temp\ipykernel_14048\1453023314.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('flight_phase').count()


,eposition,flight_datetime,eng_install_date,cycles,egt_probe_average,fuel_flw,corrected_fan_spd,core_spd,zpn12p,vib_n1_#1_bearing,vib_n2_#1_bearing,vib_n2_turbine_frame,eng_type,eng_number
flight_phase,,,,,,,,,,,,,,
CLIMB,37730,37730,37730,37725,37691,37684,0,37696,37696,37539,37538,37539,37730,37730
CRUISE,45057,45057,45057,45057,45039,45039,42192,45057,45057,45008,45008,45008,45057,45057
TAKEOFF,37890,37890,37890,37890,37880,37880,37780,37890,37890,37832,37831,37832,37890,37890


### Initial Notes:  
1- We have 3 phases.  
2- From knowledge of the business domain, the critical phases(most stresses) that determine engine life are takeoff and climb.  
3- Also from bussiness domain, each takeoff is considered one running cycle that counts into the engine remaining useful life.  
4- Flight are ordered new to old (last flight of the engine is the first in top).  

In [16]:
# show the null percentage in each column
df.isnull().sum()/len(df)*100

eposition                0.000000
flight_phase             0.000000
flight_datetime          0.000000
eng_install_date         0.000000
cycles                   0.004143
egt_probe_average        0.055520
fuel_flw                 0.061321
corrected_fan_spd       33.730537
core_spd                 0.028174
zpn12p                   0.028174
vib_n1_#1_bearing        0.246940
vib_n2_#1_bearing        0.248597
vib_n2_turbine_frame     0.246940
eng_type                 0.000000
eng_number               0.000000
dtype: float64

In [17]:
# create two cloumns date and time from the flight_datetime, the current format is '25-12-2024  15:05:00 PM'
df['flight_datetime_c'] = pd.to_datetime(df['flight_datetime'], format='%d-%m-%y %H:%M', dayfirst=True)
df['date'] = df['flight_datetime_c'].dt.date
df['time'] = df['flight_datetime_c'].dt.time
df.head(30)


,eposition,flight_phase,flight_datetime,eng_install_date,cycles,egt_probe_average,fuel_flw,corrected_fan_spd,core_spd,zpn12p,vib_n1_#1_bearing,vib_n2_#1_bearing,vib_n2_turbine_frame,eng_type,eng_number,flight_datetime_c,date,time
0,1,CRUISE,29-11-23 8:56,24-08-23 23:00,334.0,793.3125,2782.0,92.2129,103.12,86.67,0.1267,0.500,0.300,1,11,2023-11-29 08:56:00,2023-11-29,08:56:00
1,1,CRUISE,07-09-23 1:10,24-08-23 23:00,32.0,792.9250,2725.0,90.0819,103.99,86.02,0.1200,0.200,0.375,1,11,2023-09-07 01:10:00,2023-09-07,01:10:00
2,1,CRUISE,19-01-24 3:09,24-08-23 23:00,532.0,757.2250,2861.0,89.0560,101.87,83.67,0.0933,0.350,0.600,1,11,2024-01-19 03:09:00,2024-01-19,03:09:00
3,1,CLIMB,25-02-24 21:58,24-08-23 23:00,675.0,905.0375,5959.0,NaN,107.42,92.15,0.1133,0.450,0.475,1,11,2024-02-25 21:58:00,2024-02-25,21:58:00
4,1,TAKEOFF,10-07-24 16:28,24-08-23 23:00,1223.0,912.3625,2911.0,85.3176,106.87,86.45,0.0600,0.850,0.900,1,11,2024-07-10 16:28:00,2024-07-10,16:28:00
5,1,CRUISE,23-02-25 23:45,24-08-23 23:00,126.0,806.4375,2650.0,92.0863,103.02,86.50,0.0000,0.700,0.600,1,11,2025-02-23 23:45:00,2025-02-23,23:45:00
6,1,CRUISE,17-03-24 8:01,24-08-23 23:00,766.0,715.6875,2504.0,88.0891,99.87,81.40,0.0400,0.125,0.200,1,11,2024-03-17 08:01:00,2024-03-17,08:01:00
7,1,CRUISE,08-12-24 12:53,24-08-23 23:00,118.0,786.9625,3005.0,92.2527,102.16,85.80,0.0000,0.500,0.625,1,11,2024-12-08 12:53:00,2024-12-08,12:53:00
8,1,TAKEOFF,09-08-24 2:11,24-08-23 23:00,39.0,983.2375,4066.0,86.9904,110.24,90.41,0.0933,0.400,0.650,1,11,2024-08-09 02:11:00,2024-08-09,02:11:00
9,1,TAKEOFF,30-06-24 9:25,24-08-23 23:00,1178.0,938.3250,3206.0,85.0595,108.41,87.59,0.0667,0.650,0.675,1,11,2024-06-30 09:25:00,2024-06-30,09:25:00


In [18]:
#sort the data by engine number and flight datetime
df = df.sort_values(by=['eng_number', 'flight_datetime_c'])
df.head(30)

,eposition,flight_phase,flight_datetime,eng_install_date,cycles,egt_probe_average,fuel_flw,corrected_fan_spd,core_spd,zpn12p,vib_n1_#1_bearing,vib_n2_#1_bearing,vib_n2_turbine_frame,eng_type,eng_number,flight_datetime_c,date,time
105607,1,TAKEOFF,29-02-20 7:09,19-01-20 22:00,37.0,810.5125,1520.0,NaN,103.19,79.21,0.1000,0.275,0.700,2,1,2020-02-29 07:09:00,2020-02-29,07:09:00
102026,1,CLIMB,29-02-20 7:10,19-01-20 22:00,38.0,851.5125,5373.0,NaN,105.20,89.15,0.1267,0.375,0.500,2,1,2020-02-29 07:10:00,2020-02-29,07:10:00
100450,1,CRUISE,29-02-20 8:41,19-01-20 22:00,38.0,754.8250,2431.0,NaN,102.94,83.01,0.1133,0.425,0.550,2,1,2020-02-29 08:41:00,2020-02-29,08:41:00
104854,1,CRUISE,29-02-20 9:34,19-01-20 22:00,38.0,577.1125,1749.0,NaN,94.61,53.87,0.0733,0.050,0.850,2,1,2020-02-29 09:34:00,2020-02-29,09:34:00
104015,1,TAKEOFF,29-02-20 11:33,19-01-20 22:00,38.0,876.5125,1476.0,82.9875,106.00,84.16,0.0600,0.225,0.600,2,1,2020-02-29 11:33:00,2020-02-29,11:33:00
108137,1,CLIMB,29-02-20 11:33,19-01-20 22:00,39.0,870.5875,5239.0,NaN,106.23,89.70,0.1267,0.150,0.450,2,1,2020-02-29 11:33:00,2020-02-29,11:33:00
104228,1,CRUISE,29-02-20 11:55,19-01-20 22:00,39.0,755.9250,2734.0,84.9793,103.48,82.65,0.1133,0.225,0.475,2,1,2020-02-29 11:55:00,2020-02-29,11:55:00
107919,1,CRUISE,29-02-20 12:25,19-01-20 22:00,39.0,744.6125,2687.0,84.5876,102.74,81.89,0.1333,0.175,0.450,2,1,2020-02-29 12:25:00,2020-02-29,12:25:00
99529,1,CLIMB,29-02-20 18:24,19-01-20 22:00,40.0,849.6125,5337.0,NaN,105.01,88.99,0.1267,0.400,0.375,2,1,2020-02-29 18:24:00,2020-02-29,18:24:00
104341,1,TAKEOFF,29-02-20 18:24,19-01-20 22:00,39.0,804.8500,1587.0,77.9703,102.90,78.20,0.1133,0.300,0.750,2,1,2020-02-29 18:24:00,2020-02-29,18:24:00


In [19]:
# drop the flight_datetime column
df = df.drop('flight_datetime', axis=1)

In [20]:
# for each engine lets assign a flight cycle for each row, flight cycle start with take off and end with cruise
# flight cycle will increase by 1 for each take off from oldest to newest date for each engine
# first split daata by engine into a list of dataframes 
# then from oldest to newsest date for each engine assign a flight cycle for each row that increaes by 1 for each take off
# then combine the dataframes into one dataframe
df['flight_cycle'] = 0
df_list = [df[df['eng_number'] == i] for i in df['eng_number'].unique()]
for i in range(len(df_list)):
    df_list[i] = df_list[i].sort_values(by='flight_datetime_c')
    flight_cycle = 0
    for j in range(len(df_list[i])):
        if df_list[i].iloc[j]['flight_phase'] == 'TAKEOFF':
            flight_cycle += 1
        df_list[i].iloc[j, df_list[i].columns.get_loc('flight_cycle')] = flight_cycle

In [21]:
# combine the dataframes
df = pd.concat(df_list)

In [22]:
df.head(20)

,eposition,flight_phase,eng_install_date,cycles,egt_probe_average,fuel_flw,corrected_fan_spd,core_spd,zpn12p,vib_n1_#1_bearing,vib_n2_#1_bearing,vib_n2_turbine_frame,eng_type,eng_number,flight_datetime_c,date,time,flight_cycle
105607,1,TAKEOFF,19-01-20 22:00,37.0,810.5125,1520.0,NaN,103.19,79.21,0.1000,0.275,0.700,2,1,2020-02-29 07:09:00,2020-02-29,07:09:00,1
102026,1,CLIMB,19-01-20 22:00,38.0,851.5125,5373.0,NaN,105.20,89.15,0.1267,0.375,0.500,2,1,2020-02-29 07:10:00,2020-02-29,07:10:00,1
100450,1,CRUISE,19-01-20 22:00,38.0,754.8250,2431.0,NaN,102.94,83.01,0.1133,0.425,0.550,2,1,2020-02-29 08:41:00,2020-02-29,08:41:00,1
104854,1,CRUISE,19-01-20 22:00,38.0,577.1125,1749.0,NaN,94.61,53.87,0.0733,0.050,0.850,2,1,2020-02-29 09:34:00,2020-02-29,09:34:00,1
104015,1,TAKEOFF,19-01-20 22:00,38.0,876.5125,1476.0,82.9875,106.00,84.16,0.0600,0.225,0.600,2,1,2020-02-29 11:33:00,2020-02-29,11:33:00,2
108137,1,CLIMB,19-01-20 22:00,39.0,870.5875,5239.0,NaN,106.23,89.70,0.1267,0.150,0.450,2,1,2020-02-29 11:33:00,2020-02-29,11:33:00,2
104228,1,CRUISE,19-01-20 22:00,39.0,755.9250,2734.0,84.9793,103.48,82.65,0.1133,0.225,0.475,2,1,2020-02-29 11:55:00,2020-02-29,11:55:00,2
107919,1,CRUISE,19-01-20 22:00,39.0,744.6125,2687.0,84.5876,102.74,81.89,0.1333,0.175,0.450,2,1,2020-02-29 12:25:00,2020-02-29,12:25:00,2
99529,1,CLIMB,19-01-20 22:00,40.0,849.6125,5337.0,NaN,105.01,88.99,0.1267,0.400,0.375,2,1,2020-02-29 18:24:00,2020-02-29,18:24:00,2
104341,1,TAKEOFF,19-01-20 22:00,39.0,804.8500,1587.0,77.9703,102.90,78.20,0.1133,0.300,0.750,2,1,2020-02-29 18:24:00,2020-02-29,18:24:00,3


In [23]:
# add a column for remaining useful life (RUL) for each engine
# RUL is the number of flight cycles remaining for the engine before it fails
# RUL = max(flight_cycle for this engine) - flight_cycle
# store the max flight cycle for each engine in a dictionary
max_flight_cycle = {}
for i in df['eng_number'].unique():
    max_flight_cycle[i] = df[df['eng_number'] == i]['flight_cycle'].max()
df['RUL'] = 0
for i in range(len(df)):
    df.iloc[i, df.columns.get_loc('RUL')] = max_flight_cycle[df.iloc[i]['eng_number']] - df.iloc[i]['flight_cycle']


In [24]:
# lets check the data
df.head(20)

,eposition,flight_phase,eng_install_date,cycles,egt_probe_average,fuel_flw,corrected_fan_spd,core_spd,zpn12p,vib_n1_#1_bearing,vib_n2_#1_bearing,vib_n2_turbine_frame,eng_type,eng_number,flight_datetime_c,date,time,flight_cycle,RUL
105607,1,TAKEOFF,19-01-20 22:00,37.0,810.5125,1520.0,NaN,103.19,79.21,0.1000,0.275,0.700,2,1,2020-02-29 07:09:00,2020-02-29,07:09:00,1,4428
102026,1,CLIMB,19-01-20 22:00,38.0,851.5125,5373.0,NaN,105.20,89.15,0.1267,0.375,0.500,2,1,2020-02-29 07:10:00,2020-02-29,07:10:00,1,4428
100450,1,CRUISE,19-01-20 22:00,38.0,754.8250,2431.0,NaN,102.94,83.01,0.1133,0.425,0.550,2,1,2020-02-29 08:41:00,2020-02-29,08:41:00,1,4428
104854,1,CRUISE,19-01-20 22:00,38.0,577.1125,1749.0,NaN,94.61,53.87,0.0733,0.050,0.850,2,1,2020-02-29 09:34:00,2020-02-29,09:34:00,1,4428
104015,1,TAKEOFF,19-01-20 22:00,38.0,876.5125,1476.0,82.9875,106.00,84.16,0.0600,0.225,0.600,2,1,2020-02-29 11:33:00,2020-02-29,11:33:00,2,4427
108137,1,CLIMB,19-01-20 22:00,39.0,870.5875,5239.0,NaN,106.23,89.70,0.1267,0.150,0.450,2,1,2020-02-29 11:33:00,2020-02-29,11:33:00,2,4427
104228,1,CRUISE,19-01-20 22:00,39.0,755.9250,2734.0,84.9793,103.48,82.65,0.1133,0.225,0.475,2,1,2020-02-29 11:55:00,2020-02-29,11:55:00,2,4427
107919,1,CRUISE,19-01-20 22:00,39.0,744.6125,2687.0,84.5876,102.74,81.89,0.1333,0.175,0.450,2,1,2020-02-29 12:25:00,2020-02-29,12:25:00,2,4427
99529,1,CLIMB,19-01-20 22:00,40.0,849.6125,5337.0,NaN,105.01,88.99,0.1267,0.400,0.375,2,1,2020-02-29 18:24:00,2020-02-29,18:24:00,2,4427
104341,1,TAKEOFF,19-01-20 22:00,39.0,804.8500,1587.0,77.9703,102.90,78.20,0.1133,0.300,0.750,2,1,2020-02-29 18:24:00,2020-02-29,18:24:00,3,4426


In [25]:
max_flight_cycle

{1: 4429,
 2: 4690,
 3: 4510,
 4: 4035,
 5: 4427,
 6: 4089,
 7: 3975,
 8: 2018,
 9: 1888,
 10: 1967,
 11: 1862}

In [26]:
# lets add  max_flight_cycle dict to the engine life dict in days
engine_life = {}
for i in df['eng_number'].unique():
    engine_life[i] = (df[df['eng_number'] == i]['date'].max() - df[df['eng_number'] == i]['date'].min()).days

# Add max_flight_cycle to engine_life
for i in engine_life.keys():
    engine_life[i] = {'Life in Days': engine_life[i], 'Life in FC': max_flight_cycle[i]}
engine_life

{1: {'Life in Days': 1736, 'Life in FC': 4429},
 2: {'Life in Days': 1648, 'Life in FC': 4690},
 3: {'Life in Days': 1830, 'Life in FC': 4510},
 4: {'Life in Days': 1797, 'Life in FC': 4035},
 5: {'Life in Days': 1715, 'Life in FC': 4427},
 6: {'Life in Days': 1536, 'Life in FC': 4089},
 7: {'Life in Days': 1468, 'Life in FC': 3975},
 8: {'Life in Days': 646, 'Life in FC': 2018},
 9: {'Life in Days': 533, 'Life in FC': 1888},
 10: {'Life in Days': 583, 'Life in FC': 1967},
 11: {'Life in Days': 564, 'Life in FC': 1862}}

In [30]:
# save the cleaned data to a csv file
df.to_csv('engines_data_cleaned.csv', index=False)


### Notes:  
1- Our 11 engines life time ranges from 1862 FC to 4690 FC, from 533 days to 1830 days. 

2- Significant difference in life between type 1 Engines (shorter life) and type 2 engines. 

3- We have the following features to analyze: flight_cycle, egt_probe_average, fuel_flw, corrected_fan_spd, core_spd, zpn12p, vib_n1_#1_bearing, vib_n2_#1_bearing, vib_n2_turbine_frame, RUL   

4- The features are for 3 phases on 11 engines.   